### ISSUE TITLE
How to use allocated memory

### ISSUE THREAD

**Issue by cellman1234** [mtrebi/memory-allocators#2]
Hi!
First off, this is an amazing article, and it has really helped me out as a beginner to C/C++.  Unfortunately, there lies my problem: as a beginner, I am rather new to how all of this works, so please forgive me if this comes off as an unsophisticated question.

With that said, the purpose of the stack allocator (or any other allocator), as I understand it, is to segment off a piece of memory in the heap, then from elsewhere in your codebase allocate pieces of that block of memory for individual use, and free it back when you are done.  Thus, one (i.e. me, the beginner) would expect that upon allocating memory via` myStack -> Allocate(64, 8)`, it would return a pointer to the memory you allocated so that you may edit it, and then free it when you are finished.

Unfortunately for me, while it does return a pointer, I find myself unable to use it as such: upon assigning it (via `allocatedMem = &myValue`), the compiler throws an error stating that `void* is not a ponter-to-object type`. (Also, assigning it would be fruitless, as then it would no longer point to the allocated memory.)

The tl;dr of it is: **how do I use memory that I allocate**?  Are there some steps that I am missing, or is my beginner-ness just getting in the way of me thinking clearly?

Thanks in advance, -cellman123

**Comment by mtrebi** [comment]
Hi Cellman123!

First of all, I am glad you liked the article and that it helped you! C/C++ can be really tricky, specially at the beginning. So, don't worry about asking questions, we all were beginners at some point and it's extremely good to ask things you don't understand. 

I think you understanding of memory allocators is good. Just to complete it, I would say that the REAL purpose of memory allocators is to allocate memory to fit the dynamic needs of your program. That is to say, data that you can't know in advance how much memory will need, so you cannot put that in the stack. In order to do this, we use the heap and memory allocations/deallocations to ask/return memory blocks. Said that, custom allocators try to improve performance of memory allocation/deallocation by working with an internal structure that better fit our data needs.

So, all allocate methods of my allocators return a pointer. You had trouble using this pointer because it's type is actually `void *`. void pointers are a little bit special because they can point to anything. This effectively means that you cannot (directly) deference them because the compiler doesn't know its type. To get the value the pointer points to is up to the _programmer to remember its type_. This means that, if you allocate some memory and you get back a void pointer, is up to you to know the real data this pointer points to so that you can access its value:

```c
void * voidPtr = myStack->Allocate(4, 4);		// You allocate your int and get the void ptr after allocation
int * intPtr = (int *) voidPtr;								// You remember it was an int, so you cast it back
int intValue = *intPtr;										// Now you can deference the int ptr
```
So, the only thing you were missing was the cast to your-data-type ptr.

Now maybe you are wondering, why void pointers? This is so annoying. And you are totally right. Even I agree with you. But, while the usage of void pointers is ugly,  dangerous and tricky it can be really useful in some situations. For example, in my case it allows me to allocate objects without having to worry about its type. (Maybe templates would be a solution here, but I did use void pointers on purpose to match the C implementation of malloc and free).

Here there's [more info](https://www.programiz.com/c-programming/c-dynamic-memory-allocation) about void pointers, casts and allocations that may help you.

Maybe it was a too long explanation to just say that you missed the cast to your type from void pointer...Hehe

Hope this helps!

Mtrebi.

**Comment by mtrebi** [comment]
Cool! Keep me updated and tell me if you need more help :+1:
